[7.0][Machine-learning][ACR-1691] Add ml-service documentation

gichan-jang opened the pull request:
Add machine learning service API guide documentation.
 - ACR: ACR-1691

Signed-off-by: gichan <gichan2.jang@samsung.com>

jaeyun-jung commented on docs/application/native/guides/machine-learning/machine-learning-service.md:
````diff
@@ -0,0 +1,76 @@
+}
+```
+
+Get the pipeline description and constrtuct this for intelligence service.
````
typo constrtuct > construct

jaeyun-jung commented on docs/application/native/guides/machine-learning/overview.md:
```diff
@@ -13,6 +13,10 @@ You can use the following machine learning features in your native applications:
 
+- [Machine Learning Service](machine-learning-service.md)
+
+  The NNStreamer Service API provides interfaces to store and fetch the pipeline description for AI application developers.
```
recommend : NNStreamer Service API > Machine Learning Service API

gichan-jang replied:
Thanks. updated!

arisha-srbd commented on docs/application/native/guides/machine-learning/machine-learning-service.md:
```diff
@@ -0,0 +1,75 @@
+
+## Prerequisites
+
+To enable your application to use Machine Learning Service API:
```
Lists should be introduced with a complete sentence. Ref: https://developers.google.com/style/lists#introductory-sentences-for-lists
```suggestion
Follow the steps below to enable your application to use Machine Learning Service API:
```

arisha-srbd commented on docs/application/native/guides/machine-learning/machine-learning-service.md:
```diff
@@ -0,0 +1,75 @@
+
+The following code snippet shows a simple example of the Machine Learning Service API.
+
+A ML developer adds a pipeline description which includes preprocessing and invoking a neural network model, and an application developer can get the registered pipeline.
```
Incorrect article usage before the word “ML”.
```suggestion
An ML developer adds a pipeline description which includes preprocessing and invoking a neural network model, and an application developer can get the registered pipeline.
```

arisha-srbd commented on docs/application/native/guides/machine-learning/machine-learning-service.md:
```diff
@@ -0,0 +1,75 @@
+
+A ML developer adds a pipeline description which includes preprocessing and invoking a neural network model, and an application developer can get the registered pipeline.
+
+1. Set the pipeline description.
```
Code snippets must be preceded by a “colon”
```suggestion
1. Set the pipeline description:
```

arisha-srbd commented on docs/application/native/guides/machine-learning/overview.md:
```diff
@@ -13,6 +13,10 @@ You can use the following machine learning features in your native applications:
+
+  The Machine Learning Service API provides interfaces to store and fetch the pipeline description for AI application developers.
+
 ## Related Information
```
Please use sentence-style capitalization for second headings. Ref: https://github.com/Samsung/tizen-docs/blob/master/styleguide/style.md#capitalization 
```suggestion
## Related information
```

arisha-srbd reviewed (commented):
@gichan-jang Hello. Hope you are doing well.
I have gone through this PR, and have found a few errors. I made my in-line correction suggestions accordingly, please incorporate the changes. I couldn't comment on one correction, I am listing that one below: 

**File: docs/application/native/guides/machine-learning/overview.md**

1.	**Line 4:** “Machine learning features help you to handle machine learning frameworks, like TensorFlow and TensorFlow-Lite, with the construction of data pipeline.”
**Suggestion:** “Machine learning features help you to handle machine learning frameworks, like TensorFlow and TensorFlow-Lite, with the construction of a data pipeline.”
Reason: An article was missing before the word "data pipeline". 

_**Note:**_ I have a concern regarding the file: docs/application/native/guides/machine-learning/machine-learning-service.md. It shows that the dependency for the Machine Learning Service is Tizen 7.0 and Higher. Do you think it'd be ok to publish information concerning Tizen 7.0 before its release? I should let you know that once this PR will be merged to the master (live) branch, the information would be published in docs.tizen.org. Let me know what you think.

arisha-srbd commented on docs/application/native/guides/machine-learning/machine-learning-service.md:
````diff
@@ -0,0 +1,75 @@
+}
+```
+
+2. Get the pipeline description and construct this for various data-streamed service.
````
Code snippets must be preceded by a “colon”, and the word "service" should be plural here. 
```suggestion
2. Get the pipeline description and construct this for various data-streamed services: 
```

gichan-jang commented:
@arisha-srbd Thanks for kind reviews.

> I have gone through this PR, and have found a few errors. I made my in-line correction suggestions accordingly, please incorporate the changes. I couldn't comment on one correction, I am listing that one below:
> 
> **File: docs/application/native/guides/machine-learning/overview.md**
> 
> 1. **Line 4:** “Machine learning features help you to handle machine learning frameworks, like TensorFlow and TensorFlow-Lite, with the construction of data pipeline.”
>    **Suggestion:** “Machine learning features help you to handle machine learning frameworks, like TensorFlow and TensorFlow-Lite, with the construction of a data pipeline.”
>    Reason: An article was missing before the word "data pipeline".
> 

Updated!

> _**Note:**_ I have a concern regarding the file: docs/application/native/guides/machine-learning/machine-learning-service.md. It shows that the dependency for the Machine Learning Service is Tizen 7.0 and Higher. Do you think it'd be ok to publish information concerning Tizen 7.0 before its release? I should let you know that once this PR will be merged to the master (live) branch, the information would be published in docs.tizen.org. Let me know what you think.

 Right. I changed base from `master` to `tizen_7.0_prepare`. Thanks.

arisha-srbd commented on docs/application/native/guides/machine-learning/machine-learning-service.md:
````diff
@@ -0,0 +1,75 @@
+}
+```
+
+2. Get the pipeline description and construct this for various data-streamed service:
````
@gichan-jang perhaps you missed this one. The correct form of the word "service" would be "services" here. Please incorporate this remaining change. Thanks. 
```suggestion
2. Get the pipeline description and construct this for various data-streamed services:
```

gichan-jang replied:
Updated. Thanks.

mijin85cho commented on docs/application/native/guides/machine-learning/overview.md:
```diff
@@ -13,7 +13,11 @@ You can use the following machine learning features in your native applications:
+## Related information
 
 - Dependencies
   - Tizen 5.5 and Higher for Mobile
```
For other sub pages, "Tizen 5.5 and Higher for Mobile/Wearable" is correct.

However in case of "Machine Learning Service", 
according to added machine-learning-service.md, "Tizen 7.0 and Higher for TV" & "Tizen 7.0 and Higher for IoT" are also supported.
As the Overview page is also covering "Machine Learning Service",
how about adding "Tizen 7.0 and Higher for TV" & "Tizen 7.0 and Higher for IoT" to the overview.md?

gichan-jang replied:
The dependency information was omitted. I have updated it, so please check again.

mijin85cho reviewed (commented):
Hello, @arisha-srbd

Thanks for the review.
I briefly checked the PR, and got below comments. Would you please include these to the review that your doing, too?

asm-foysal commented on docs/application/native/guides/machine-learning/machine-learning-service.md:
```diff
@@ -0,0 +1,75 @@
+# Machine Learning Service
```
@mijin85cho hello!
We go through the checklist provided by SRIB before approving a PR.
We did notice that the file was missing from the LNB, Arisha was planning on providing another review and mentioning the LNB in that one.
Apologies for missing this in the first review, we'll try and tighten up our review process further.

mijin85cho replied:
Hello, @asm-foysal !
Thanks for checking & reply. I was just worrying about the case that point get missed.
If you were about to point out also, my concern is resolved.

gichan-jang replied:
Updated. PTAL.

gichan-jang commented on docs/application/native/guides/machine-learning/machine-learning-service.md:
````diff
@@ -0,0 +1,75 @@
+An ML developer adds a pipeline description which includes preprocessing and invoking a neural network model, and an application developer can get the registered pipeline.
+
+1. Set the pipeline description:
+```c
````
Updated!

arisha-srbd reviewed (commented):
@gichan-jang hello. Hope you are doing well. Thank you for incorporating all the previous issues. I made a correction suggestion in the toc_all.md file. Please make the change. 

I also reviewed the newly added files to this PR and there are a few issues that I found. Please incorporate the changes that I am listing below: 

_**File: docs/application/native/guides/machine-learning/machine-learning-inference.md**_

1.	Line 6: “The main features of the Machine Learning Inference API include:”
Suggestion: “The main features of the Machine Learning Inference API include the following:”
Reason: Lists should be introduced with a complete sentence. Ref: https://developers.google.com/style/lists#introductory-sentences-for-lists

2.	Line 24: “Custom filters, which are neural network models implemented directly with programming languages including C/C++ and Python, maybe imported as filters of pipelines directly as well.”
Suggestion: “Custom Filters, which are neural network models implemented directly with programming languages including C/C++ and Python, maybe imported as filters of pipelines directly as well.”
Reason: In all other instances of this file, the word “Custom Filter” has been treated as a proper noun and capitalized.  

3.	Line 31: “To enable your application to use the machine learning functionality:”
Suggestion: “To enable your application to use the machine learning functionality, go through the following steps:”
Reason: Lists should be introduced with a complete sentence. Ref: https://developers.google.com/style/lists#introductory-sentences-for-lists

4.	Line 31: “This section shows how to load a model without the construction of pipelines.”
Suggestion: “This section shows how to load a model without the construction of pipelines:”
Reason: Ordered lists should be preceded by a colon. 

5.	Line 63: “To load a model file, two `ml_tensors_info_h` are required. `in_info` contains the information of the input tensors, and `out_info` contains the information of the output tensors. For more information, see [Tensors Information](#tensors-information).”

_The line talks about two `ml_tensors_info_h`, however the code above this section doesn’t mention anything about “two”. Please look into this if it would be an issue._ 

6.	Line 93: “Basic Usage”
Suggestion: “Basic usage”
Reason: Please use sentence style capitalization for third headings. Ref: https://github.com/Samsung/tizen-docs/blob/master/styleguide/style.md#capitalization 

7.	Line 133: “Followings are the available elements:”
Suggestion: “Following are the available elements:”
Reason: The word “following” should be singular here. 

8.	 Line 468: “Pipeline States”
Suggestion: “Pipeline states”
Reason: Please use sentence style capitalization for third headings. Ref: https://github.com/Samsung/tizen-docs/blob/master/styleguide/style.md#capitalization 

9.	Line 472: “Tensors Information”
Suggestion: “Tensors information”
Reason: Please use sentence style capitalization for second headings. Ref: https://github.com/Samsung/tizen-docs/blob/master/styleguide/style.md#capitalization 

10.	Line 520: “Tensors Data”
Suggestion: “Tensors data”
Reason: Please use sentence style capitalization for second headings. Ref: https://github.com/Samsung/tizen-docs/blob/master/styleguide/style.md#capitalization 

11.	Line 550: “….Since the callback function works as **filter** in the pipeline, it is named as `Custom Filter`.”
Suggestion: “….Since the callback function works as a **filter** in the pipeline, it is named `Custom Filter`.”
Reason: Missing article before the word “filter”. 

12.	Line 614: “Related Information”
Suggestion: “Related information”
Reason: Please use sentence style capitalization for second headings. Ref: https://github.com/Samsung/tizen-docs/blob/master/styleguide/style.md#capitalization 

_**File: docs/application/native/guides/machine-learning/machine-learning-training.md**_

1.	Line 7: “Constructing deep neural network (DNN)”
Suggestion: “Constructing a deep neural network (DNN)”
Reason: An article before the word “deep” might be appropriate here, please look into this. 

2.	Line 16: “Error must be handled more extensively compared to the example code written in this page.” 
Suggestion: “The error must be handled more extensively compared to the example code written on this page.”
Reason: An article missing before the word “error” and preposition usage problem. 

3.	Line 20: “To enable your application to use Machine Learning Training API:”
Suggestion: “To enable your application to use Machine Learning Training API, go through the following steps:”
Reason: Lists should be introduced with a complete sentence. Ref: https://developers.google.com/style/lists#introductory-sentences-for-lists

4.	Line 35: _**Is this paragraph a part of the ordered list? If so then it should be indented accordingly.**_ 

5.	Line 112: “One type includes commonly trainable weights and the other type does not include.”
Suggestion: “One type includes commonly trainable weights and the other type does not include them.”
Reason: The line seems incomplete if we end it with “include”. 

6.	Line 117: “Universal properties that applies to every layer”
Suggestion: “Universal properties that apply to every layer”
Reason: The word “apply” should be singular here. 

7.	Line 120: “Comma-seperated names of layers to be inputs of the current layer”
Suggestion: “Comma-separated names of layers to be inputs of the current layer”
Reason: Spelling correction. 

8.	Line 121: “Comma-seperated Formatted string as "channel:height:width". If there is no channel then it must be 1. First layer of the model must have input_shape. Other can be omitted as it is calculated at compile phase.”
Suggestion: “Comma-separated formatted string as "channel:height:width". If there is no channel then it must be 1. The first layer of the model must have input_shape. Other can be omitted as it is calculated at compile phase.”
Reason: Should the word “formatted” be capitalized here? An article missing before the word “first”. 

9.	Line 152: “Comma-seperated unsigned integers for kernel size, `height, width` respectively”
Suggestion: “Comma-separated unsigned integers for kernel size, `height, width` respectively”

10.	Line 153: “Comma-seperated unsigned integers for strides, `height, width` respectively”
Suggestion: “Comma-separated unsigned integers for strides, `height, width` respectively”

11.	Line 162: “Word embeddeing size”
Suggestion: “Word embedding size”
Reason: Spelling correction. 

12.	Line 201: “The following are the available properties for each layer type which does not include (`weight_initializer`, `bias_initializer`, `weight_regularizer`, `weight_regularizer_constant`) properties.”
Suggestion: “The following are the available properties for each layer type which do not include (`weight_initializer`, `bias_initializer`, `weight_regularizer`, `weight_regularizer_constant`) properties.”
Reason: The correct verb form to be used here is “do”. 

13.	Line 205: “Universal properties that applies to every layer”
Suggestion: “Universal properties that apply to every layer”
Reason: The word “apply” should be singular here. 

14.	Line 208: “Comma-seperated names of layers to be inputs of the current layer”
Suggestion: “Comma-separated names of layers to be inputs of the current layer”
Reason: Spelling correction. 

15.	Line 209: “Comma-seperated Formatted string as "channel:height:width". If there is no channel then it must be 1. First layer of the model must have input_shape. Other can be omitted as it is calculated at compile phase.”
Suggestion: “Comma-separated formatted string as "channel:height:width". If there is no channel then it must be 1. The first layer of the model must have input_shape. Other can be omitted as it is calculated at compile phase.”
Reason: Should the word “formatted” be capitalized here? An article missing before the word “first”. 

16.	Line 221: “Small value to avoid divide by zero”
Suggestion: “Small value to avoid dividing by zero”
Reason: The correct verb form to be used is “dividing”. 

17.	Line 261: “Comma-seperated unsigned integers for pooling size, `height, width` respectively”
Suggestion: “Comma-separated unsigned integers for pooling size, `height, width` respectively”

18.	Line 262: “Comma-seperated unsigned integers for stride, `height, width` respectively”
Suggestion: “Comma-separated unsigned integers for stride, `height, width` respectively”

19.	Line 291: “Vertiacl direction”
Suggestion: “Vertical direction”
Reason: Spelling correction. 

20.	Line 310: “Configure a optimizer”
Suggestion: “Configure an optimizer”
Reason: The correct article to be used before “optimizer” 

21.	Line 330: “beta1 coefficient for adam”
Suggestion: “Beta1 coefficient for Adam”
Reason: To maintain consistency with the entire table's content, the first word and “Adam” should be capitalized. 

22.	Line 331: “beta2 coefficient for adam”
Suggestion: “Beta2 coefficient for Adam”
Reason: To maintain consistency with the entire table's content, the first word and “Adam” should be capitalized. 

23.	Line 332: “epsilon coefficient for adam”
Suggestion: “Epsilon coefficient for Adam”
Reason: To maintain consistency with the entire table's content, the first word and “Adam” should be capitalized. 

24.	Line 459: “Example code written here reproduces the model description from [Create Model from INI Formatted File](#create-a-model-from-ini-formatted-file) except that followings are different:”
Suggestion: “Example code written here reproduces the model description from [Create Model from INI Formatted File](#create-a-model-from-ini-formatted-file) except that following are different:”
Reason: The word "following" should be singular here. 

25.	Line 539: “A batch is a bundle of samples which is fed to a single iteration. If a batch is of size 5, the batch layout, where `/*` denotes nth sample will be:”
Suggestion: “A batch is a bundle of samples which is fed to a single iteration. If a batch is of size 5, the batch layout, where `/*` denotes the nth sample will be as follows:”
Reason: The code snippet should be introduced by a complete sentence. 

26.	Line 553-554: “So, if you consider [Cifar-100](https://www.cs.toronto.edu/~kriz/cifar.html#The%20CIFAR-100%20dataset) dataset.
When using the full training set, a single training epoch will contain 50,000 samples. If batch size is 100, it will be of 500 batches(or iterations).”
Suggestion: “So, if you consider [Cifar-100](https://www.cs.toronto.edu/~kriz/cifar.html#The%20CIFAR-100%20dataset) dataset
when using the full training set, a single training epoch will contain 50,000 samples. If batch size is 100, it will be of 500 batches (or iterations).”
Reason: If line 553 is kept as it is, it sounds incomplete. So, I think it should be joined with line 554. 

27.	Line 568: “To create a `dataset` from files. Each file must contains an array of samples.”
Suggestion: “To create a `dataset` from files, each file must contain an array of samples.”
Reason: The two sentences should be joined to make it understandable. The word “contain” should be singular here. 

28.	Line 569: “A single sample consists of an array of raw float array for input and another array or raw float array for labels which contains same size as model output:”
Suggestion: “A single sample consists of an array of raw float array for input and another array or raw float array for labels that contains same size as model output:”

29.	Line 572: “If a model requires to two inputs and a single label for a sample, a single sample would contain...”
Suggestion: “If a model requires two inputs and a single label for a sample, a single sample would contain...”
Reason: I think the preposition “to” is not necessary here. 

30.	Line 579: “In Cifar-100 data, it contains a image of size 3 * 32 * 32 and a coarse label of 20 classes and a fine label of 100 classes.”
Suggestion: “In Cifar-100 data, it contains an image of size 3 * 32 * 32 and a coarse label of 20 classes and a fine label of 100 classes.”
Reason: Incorrect article usage. 

31.	Line 589: “If your model requires, an image and two labels (coarse, fine) both one-hot encoded for a sample, the file layout would be:”
Suggestion: “If your model requires an image and two labels (coarse, fine) both one-hot encoded for a sample, the file layout would be:
Reason: Removed comma after the word “requires” 

32.	Line 694: “For example, you can use the trained model with single API as follows:”
Suggestion: “For example, you can use the trained model with a single API as follows:”
Reason: An article missing before the word “single”.

gichan-jang commented on docs/application/toc_all.md:
```diff
@@ -882,6 +882,7 @@
 #### [Overview](/application/native/guides/machine-learning/overview.md)
 #### [Machine Learning Inference](/application/native/guides/machine-learning/machine-learning-inference.md)
 #### [Machine Learning Training](/application/native/guides/machine-learning/machine-learning-training.md)
+#### [Machine Learning Training](/application/native/guides/machine-learning/machine-learning-service.md)
```
My bad. Updated.

gichan-jang commented:
@arisha-srbd Thank you for your kind review and guide.
I updated the documentation. Please review it again. Thanks.

arisha-srbd commented on docs/application/native/guides/machine-learning/machine-learning-training.md:
```diff
@@ -565,20 +563,20 @@ Consider setting property with `ml_train_model_set_property_for_mode()` when dat
 
-To create a `dataset` from files. Each file must contains an array of samples.
-A single sample consists of an array of raw float array for input and another array or raw float array for labels which contains same size as model output:
+To create a `dataset` from files. Each file must contain an array of samples.
```
If the two sentences are joined, it will adequately make sense. Please let me know what you think. 
```suggestion
To create a `dataset` from files, each file must contain an array of samples.
```

arisha-srbd commented on docs/application/native/guides/machine-learning/machine-learning-training.md:
```diff
@@ -198,15 +197,15 @@ Type | Key | Value | Default value | Description
 &#xfeff;                                                     | dropout                     | (float)                     | 0                       | Dropout rate
 
-The following are the available properties for each layer type which does not include (`weight_initializer`, `bias_initializer`, `weight_regularizer`, `weight_regularizer_constant`) properties.
+The following are the available properties for each layer type which do not include (`weight_initializer`, `bias_initializer`, `weight_regularizer`, `weight_regularizer_constant`) properties.
```
To maintain consistency with the rest of the content, a colon should be introduced here at the end of the sentence. 
```suggestion
The following are the available properties for each layer type which do not include (`weight_initializer`, `bias_initializer`, `weight_regularizer`, `weight_regularizer_constant`) properties:
```

arisha-srbd commented on docs/application/native/guides/machine-learning/machine-learning-training.md:
```diff
@@ -536,7 +535,7 @@ A sample denotes a single pair of inputs/labels.
 2. Batch
 
-A batch is a bundle of samples which is fed to a single iteration. If a batch is of size 5, the batch layout, where `/*` denotes nth sample will be:
+A batch is a bundle of samples which is fed to a single iteration. If a batch is of size 5, the batch layout, where `/*` denotes nth sample will be as follows:
```
**Line 538-546:** Proper indentation should be maintained under the bullet points.

gichan-jang replied:
Updated

arisha-srbd commented on docs/application/native/guides/machine-learning/machine-learning-training.md:
```diff
@@ -550,8 +549,7 @@ A batch is a bundle of samples which is fed to a single iteration. If a batch is
 
 An epoch refers to a full, exhaustive visit to a dataset.
```
**Line 550-552:** Proper indentation should be maintained under the bullet points.

arisha-srbd reviewed (commented):
@gichan-jang thank you for making all the changes very promptly. I made in-line comments to make a few minor fixes. Please address those accordingly. Thank you. 

For the ones where I could not comment, I am listing those below: 

**File: docs/application/native/guides/machine-learning/machine-learning-training.md**

1.	**Line 526:** “Let's assume a given model requires three inputs and two labels.”
Suggestion: “Let's assume a given model requires three inputs and two labels. This has been reflected in the examples below: ”
Reason: Lists should be introduced with a complete sentence. Ref: https://developers.google.com/style/lists#introductory-sentences-for-lists

2.	**Line 530-534:** Proper indentation should be maintained under the bullet points.

gichan-jang commented on docs/application/native/guides/machine-learning/machine-learning-inference.md:
````diff
@@ -60,7 +60,7 @@ This section shows how to load a model without the construction of pipelines.
     ```
 
-    To load a model file, two `ml_tensors_info_h` are required. `in_info` contains the information of the input tensors, and `out_info` contains the information of the output tensors. For more information, see [Tensors Information](#tensors-information).
+    If the given model has flexible input/output dimension, `in_info` and `out_info` are required. `in_info` contains the information of the input tensors, and `out_info` contains the information of the output tensors. You may set NULL if it's not required. For more information, see [Tensors Information](#tensors-information).
````
Updated.

gichan-jang reviewed (commented):
@arisha-srbd , Thank you for your review. I updated docs.

arisha-srbd commented on docs/application/native/guides/machine-learning/machine-learning-training.md:
```diff
@@ -524,34 +523,33 @@ In either case, you need to provide streams of tensor data and arrays of values
 1. Sample
 
-A sample denotes a single pair of inputs/labels.
+    A sample denotes a single pair of inputs/labels.
```
@gichan-jang, please check this suggestion. Code snippets should be preceded by an introductory line ending with a colon. 
```suggestion
    A sample denotes a single pair of inputs/labels. The example can be as follows: 
```

gichan-jang replied:
Right. Updated.
